core/src/test/java/org/apache/calcite/test/SqlHintsConverterTest.java

```diff
@@ -578,6 +579,21 @@ public final Fixture sql(String sql) {
         .check();
   }
 
+  @Test void testHintExcludeRules() {
+    final String sql = "select empno from (select * from "
+            + "(select /*+ preserved_project */ empno, ename, deptno from emp)"
+            + " where deptno = 20)";
+
+    final RelNode rel = ruleFixture().sql(sql).toRel();
+    HepProgram program = new HepProgramBuilder()
+            .addRuleInstance(CoreRules.FILTER_PROJECT_TRANSPOSE)
+            .build();
+    HepPlanner planner = new HepPlanner(program);
+    planner.setRoot(rel);
+    RelNode newRel = planner.findBestExp();
+    Assertions.assertTrue(rel.deepEquals(newRel));
```

Comment:
It's convenient to show expected vs computed when test fails, otherwise you need to add prints or open the debugger to see what's going on.

Consider something along this line:
```suggestion
    Assertions.assertTrue(rel.deepEquals(newRel),
        "Expected:\n" + RelOptUtil.toString(rel) + "Computed:\n" + RelOptUtil.toString(newRel));
```

Alternatively, you might compare the strings directly if possible, the diff would be even easier to inspect.